File: plugins/inc.py
import re

def beeld_plugin_inc(beelder):
    beelder.addAction('inc', beeld_plugin_action_inc)

def beeld_plugin_action_inc(evt):
    params = evt.data
    data = params.data
    current = params.current
    inc = current.action_cfg
    if inc:
        p = str(inc['prefix'] if 'prefix' in inc else '')
        s = str(inc['suffix'] if 'suffix' in inc else '')
        defined = list(map(lambda i: i.strip(), str(inc['define'] if 'define' in inc else '').split(',')))
        rex = re.compile(r'^' + re.escape(p) + r'#(ifn?def)\(([^\(\)]+)\)' + re.escape(s), re.M)
        e = p + '#endif' + s
        el = len(e)
        src = data.src
        slen = len(src)
        out = ''
        offset = 0
        # extract conditional include blocks
        m = rex.search(src, offset)
        while m:
            out += src[offset:m.start(0)]
            start = m.start(0) + len(m.group(0))
            end = src.find(e, start)
            offset = slen if -1 == end else end+el
            if -1 == end: end = slen
            param = m.group(2).strip()
            if (('ifndef' == m.group(1)) and (param not in defined)) or (('ifdef' == m.group(1)) and (param in defined)): out += src[start:end]
            m = rex.search(src, offset)

        if offset < slen: out += src[offset:slen]
        data.src = out

    evt.next()

File: plugins/inc.php
<?php
function beeld_plugin_inc($beelder)
{
    $beelder->addAction('inc', 'beeld_plugin_action_inc');
}
function beeld_plugin_action_inc($evt)
{
    $params =& $evt->data;
    $data =& $params->data;
    $current =& $params->current;
    $inc = $current->action_cfg;
    if (!empty($inc))
    {
        $p = (string)(isset($inc['prefix']) ? $inc['prefix'] : '');
        $s = (string)(isset($inc['suffix']) ? $inc['suffix'] : '');
        $defined = array_map('trim', explode(',', (string)(isset($inc['define']) ? $inc['define'] : '')));
        $rex = '/^' . preg_quote($p, '/') . '#(ifn?def)\\(([^\\(\\)]+)\\)' . preg_quote($s, '/') . '/um';
        $e = $p . '#endif' . $s;
        $el = strlen($e);
        $src = $data->src;
        $slen = strlen($src);
        $out = '';
        $offset = 0;
        // extract conditional include blocks
        while (preg_match($rex, $src, $m, PREG_OFFSET_CAPTURE, $offset))
        {
            $out .= substr($src, $offset, $m[0][1]-$offset);
            $start = $m[0][1] + strlen($m[0][0]);
            $end = strpos($src, $e, $start);
            $offset = false === $end ? $slen : $end+$el;
            if (false === $end) $end = $slen;
            $param = trim($m[2][0]);
            if (
                ('ifndef' === $m[1][0] && !in_array($param, $defined)) ||
                ('ifdef' === $m[1][0] && in_array($param, $defined))
            ) $out .= substr($src, $start, $end-$start);
        }
        if ($offset < $slen) $out .= substr($src, $offset, $slen-$offset);
        $data->src = $out;
    }
    $evt->next();
}

